### PR TITLE
add latency and availability metrics to gateway

### PR DESCRIPTION
**Pull request by zeeshan-mohd**
Added metrics that needs to be sent for latency ( in case of middleware request and response ) and availability metrics for middlewares requests

### CODE REVIEW

File: runtime/middlewares.go [hunk 1]
```diff
@@ -103,24 +110,59 @@ func (m *MiddlewareStack) Handle(
...
+
+			start = time.Now() // start the timer for middleware responses
 			for j := i; j >= 0; j-- {
 				m.middlewares[j].HandleResponse(ctx, res, shared)
 			}
+			//record latency for middlewares responses in unsuccesful case
+			m.recordLatency(middlewareResponseLatencyTag, start, req)
+
+			//for error metrics only emit when there is gateway error and not request error
+			if res.pendingStatusCode >= 500 {
```
Comment (isopropylcyanide) [inline runtime/middlewares.go:132]: 4xx is a success?

File: runtime/middlewares.go [hunk 2]
```diff
@@ -22,10 +22,17 @@ package zanzibar
 
 import (
 	"context"
+	"time"
 
 	jsonschema "github.com/mcuadros/go-jsonschema-generator"
 )
 
+const (
+	middlewareRequestLatencyTag  = "middleware.requests.latency"
```
Comment (isopropylcyanide) [inline runtime/middlewares.go:31]: Should be singular `middleware.request.latency`

File: runtime/middlewares.go [hunk 3]
```diff
@@ -103,24 +110,59 @@ func (m *MiddlewareStack) Handle(
...
 		}
 	}
+	// record latency for middlewares requests in successful case
+	m.recordLatency(middlewareRequestLatencyTag, start, req)
+	//emit success metric for middleware requests
+	m.emitAvailability(middlewareRequestStatusTag, "success", req)
 
 	ctx = m.handle(ctx, req, res)
 
+	start = time.Now()
```
Comment (bishnuag) [inline runtime/middlewares.go:147]: multiple usage of start var, lets rename to middlewareResponseStartTIme

File: runtime/middlewares.go [hunk 4]
```diff
@@ -103,24 +110,59 @@ func (m *MiddlewareStack) Handle(
...
 }
+
+// recordLatency measures the latency as per the tagName and start time given.
+func (m *MiddlewareStack) recordLatency(tagName string, start time.Time, req *ServerHTTPRequest) {
+	elapsed := time.Now().Sub(start)
+	req.scope.Timer(tagName).Record(elapsed)
+}
+
+// emitAvailability is used to increment the success/error counter for a particular tagName.
+func (m *MiddlewareStack) emitAvailability(tagName string, status string, req *ServerHTTPRequest) {
```
Comment (bishnuag) [inline runtime/middlewares.go:163]: pass only req.scope instead entire request param

File: runtime/middlewares.go (hunk 1 above)
Comment (bishnuag) [reply runtime/middlewares.go:132]: I think it's the wrong place to verify statusCode. statusCode would be available only after we make call to downstream client - **ctx = m.handle(ctx, req, res)**

Comment (zeeshan-mohd) [reply runtime/middlewares.go:134]: > 4xx is a success?

Removed the success part as it can be taken from total_requests.Not taking 4xx as error is because the availability defined for gateway should not be dependent on the user's request and rather on the gateway internal errors, so that is why only 5xx.

**Review by bishnuag — APPROVED** [review]
LGTM

File: runtime/middlewares.go (hunk 1 above)
Comment (zeeshan-mohd) [reply runtime/middlewares.go:134]: > I think it's the wrong place to verify statusCode. statusCode would be available only after we make call to downstream client - **ctx = m.handle(ctx, req, res)**

https://github.com/uber/zanzibar/blob/24bea8e47a8464853e5014da44ac30c2b17f091b/runtime/server_http_response.go#L271
We are setting the pendingStatuscode whenever we call the res.WriteJson from the middlewares,so this will always be populated in case of errors.

File: runtime/middlewares.go [hunk 5]
```diff
@@ -103,24 +111,57 @@ func (m *MiddlewareStack) Handle(
...
 	shared := NewSharedState(m.middlewares)
+	middlewareRequestStartTime := time.Now()
 
 	for i := 0; i < len(m.middlewares); i++ {
 		ctx, ok := m.middlewares[i].HandleRequest(ctx, req, res, shared)
 		// If a middleware errors and writes to the response header
 		// then abort the rest of the stack and evaluate the response
 		// handlers for the middlewares seen so far.
 		if ok == false {
+			//record latency for middlewares requests in unsuccesful case as the middleware requests calls are terminated
```
Comment (isopropylcyanide) [inline runtime/middlewares.go:122]:
```suggestion
			//record latency for middlewares requests in unsuccessful case as the middleware requests calls are terminated
```

**Comment by isopropylcyanide** [comment]
Please add the test plan and a dump of the new metrics from a sample Zanzibar example gateway for both successful / unsuccessful runs.

File: runtime/middlewares_test.go [hunk 6]
```diff
@@ -183,7 +184,7 @@ func TestMiddlewareRequestAbort(t *testing.T) {
 	if !assert.NoError(t, err) {
 		return
 	}
-	assert.Equal(t, resp.StatusCode, http.StatusOK)
+	assert.Equal(t, resp.StatusCode, http.StatusInternalServerError)
```
Comment (isopropylcyanide) [inline runtime/middlewares_test.go:187]: Why is this changed?

**Review by isopropylcyanide — CHANGES_REQUESTED** [review]
Gave some comments

File: runtime/middlewares.go [hunk 7]
```diff
@@ -103,24 +111,57 @@ func (m *MiddlewareStack) Handle(
...
+			m.recordLatency(middlewareRequestLatencyTag, middlewareRequestStartTime, req.scope)
+
+			middlewareResponseStartTime := time.Now() // start the timer for middleware responses
 			for j := i; j >= 0; j-- {
 				m.middlewares[j].HandleResponse(ctx, res, shared)
 			}
+			//record latency for middlewares responses in unsuccesful case
+			m.recordLatency(middlewareResponseLatencyTag, middlewareResponseStartTime, req.scope)
```
Comment (isopropylcyanide) [inline runtime/middlewares.go:130]: Are we tracking req/resp latency for each middleware or all of the edge middlewares?

> If it's the former, I don't think the code does what it is supposed to do

If there are 4 middlewares and handling fails for the 4th one, then we need to run the decrement loop of handling th responses for the 3 that executed. 
In this case, the middleware response latency recorded for the 3rd would be incorrect as it will include the latency of the first 2 middlewares's responses.

```
for j := i; j >= 0; j-- {
				m.middlewares[j].HandleResponse(ctx, res, shared)
			}
m.recordLatency(middlewareResponseLatencyTag, middlewareResponseStartTime, req.scope)
```

**Review by isopropylcyanide — COMMENTED** [review]
Requested clarification

File: runtime/middlewares_test.go (hunk 6 above)
Comment (zeeshan-mohd) [reply runtime/middlewares_test.go:187]: It was changed because if we are reaching here means that the middleware request has aborted with an error and we will return from here without calling the downstream .So sending a 200 ok in this case is not apt.Additionally , it increased the code coverage as well.

File: runtime/middlewares.go (hunk 5 above)
Comment (sachsingh) [reply runtime/middlewares.go]: updated

File: runtime/middlewares.go (hunk 7 above)
Comment (zeeshan-mohd) [reply runtime/middlewares.go:130]: > Are we tracking req/resp latency for each middleware or all of the edge middlewares?

Yes , it is for all the middlewares that a request goes through , since these metrics are tagged with endpointID we will apply the filtering . So we are not measuring for one middleware but rather all the middleware that a request goes through and filter on the basis of endpointID's.

**Comment by sachsingh** [comment]
> Please add the test plan and a dump of the new metrics from a sample Zanzibar example gateway for both successful / unsuccessful runs.

Added metrics dump https://code.uberinternal.com/P341907

**Review by isopropylcyanide — APPROVED** [review]
